src/infrasys/chronify_time_series_storage.py

```diff
@@ -0,0 +1,283 @@
+        # Manage a mapping of UUIDs to integer IDs until we can remove UUIDs.
+        self._uuid_lookup: dict[UUID, int] = uuid_lookup or {}
+        self._id_manager = id_manager
+
+    @classmethod
+    def create_with_temp_directory(
+        cls,
+        base_directory: Optional[Path] = None,
+        engine_name: str = "duckdb",
+        read_only: bool = False,
+    ) -> Self:
+        """Construct ChronifyTimeSeriesStorage with a temporary directory."""
+        with NamedTemporaryFile(dir=base_directory, suffix=".db") as f:
+            dst_file = Path(f.name)
+        logger.debug("Creating database at {}", dst_file)
+        atexit.register(delete_if_exists, dst_file)
+        store = Store(engine_name=engine_name, file_path=dst_file)
+        id_manager = IDManager(next_id=1)
+        return cls(store, id_manager, read_only=read_only)
+
+    @classmethod
+    def from_file_to_tmp_file(
+        cls,
+        data: dict[str, Any],
+        dst_dir: Optional[Path] = None,
+        read_only: bool = False,
+    ) -> Self:
+        """Construct ChronifyTimeSeriesStorage after copying from an existing database file."""
+        id_manager, uuid_lookup = cls._deserialize_ids(data)
+        with NamedTemporaryFile(dir=dst_dir, suffix=".db") as f:
+            dst_file = Path(f.name)
+        orig_store = Store(engine_name=data["engine_name"], file_path=data["filename"])
+        orig_store.backup(dst_file)
+        new_store = Store(engine_name=data["engine_name"], file_path=dst_file)
+        atexit.register(delete_if_exists, dst_file)
+        return cls(new_store, id_manager, read_only=read_only, uuid_lookup=uuid_lookup)
+
+    @classmethod
+    def from_file(cls, data: dict[str, Any], read_only: bool = False) -> Self:
+        """Construct ChronifyTimeSeriesStorage with an existing database file."""
+        id_manager, uuid_lookup = cls._deserialize_ids(data)
+        store = Store(engine_name=data["engine_name"], file_path=Path(data["filename"]))
+        return cls(store, id_manager, read_only=read_only, uuid_lookup=uuid_lookup)
+
+    @staticmethod
+    def _deserialize_ids(data: dict[str, Any]) -> tuple[IDManager, dict[UUID, int]]:
+        uuid_lookup: dict[UUID, int] = {}
+        max_id = 0
+        for key, val in data["uuid_lookup"].items():
+            uuid_lookup[UUID(key)] = val
+            if val > max_id:
+                max_id = val
+        id_manager = IDManager(next_id=max_id + 1)
+        return id_manager, uuid_lookup
+
+    def get_database_url(self) -> str:
+        """Return the path to the underlying database."""
+        assert self._store.engine.url.database is not None
+        # We don't expect to use an in-memory db.
+        return self._store.engine.url.database
+
+    def get_time_series_directory(self) -> Path:
+        assert self._store.engine.url.database is not None
+        return Path(self._store.engine.url.database).parent
+
+    def add_time_series(
```

Comment:
Adding multiple time series will get exponentially slower the more you add. This is bound to be problematic. I could at least add a bulk-add procedure that only performs the checks at the end. Should be done in this PR. I'll revert to draft state.